keyhub/resource_vaultrecord.go

```diff
@@ -19,7 +20,22 @@ func resourceVaultRecord() *schema.Resource {
 		UpdateContext: resourceVaultRecordUpdate,
 		DeleteContext: resourceVaultRecordDelete,
 		Schema:        VaultRecordResourceSchema(),
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceVaultRecordImportContext,
+		},
+	}
+}
+
+func resourceVaultRecordImportContext(ctx context.Context, d *schema.ResourceData, m interface{}) ([]*schema.ResourceData, error) {
+
+	grpUuid, err := uuid.Parse(d.Id())
+	if err != nil {
+		return nil, fmt.Errorf("`%s` is not a valid uuid", d.Id())
 	}
+
+	d.Set("uuid", grpUuid.String())
```

Comment:
Dit kan niet. `d.Id()` is geen UUID maar het database ID.
Zie:
- [Resource Create](https://github.com/topicuskeyhub/terraform-provider-keyhub/blob/5674358ad4195db19f37b0b1c4fc70d67b3cbf79/keyhub/resource_vaultrecord.go#L161)
- [Data Read](https://github.com/topicuskeyhub/terraform-provider-keyhub/blob/5674358ad4195db19f37b0b1c4fc70d67b3cbf79/keyhub/data_source_vaultrecord.go#L307)

Reply:
Waar zie jij `d.Id()` gezet worden dan? :) 

De import bereid het object voor, waarna het langs de read functie gaat om werkelijk te importeren.
`terraform state import data.keyhub_vaultrecord.data2  "407532c4-fa95-4588-abd4-5a156a802efd" `
doet onderwater het zelfde als 
```
data "keyhub_vaultrecord" "data2" {
  uuid = "407532c4-fa95-4588-abd4-5a156a802efd"
}
```

Demo:
```hcl
resource "keyhub_vaultrecord" "record" {
  groupuuid = "4c25e8df-9bba-4e63-886c-b173d62fac55"
  name      = "Terraform Test record"
  username  = "a_username_s"
  password  = "just_a_pass"
}

```

```
% terraform state show keyhub_vaultrecord.record
No state file was found!

State management commands require a state file. Run this command
in a directory where Terraform has been run or use the -state flag
to point the command to a specific state location.
```

```
% terraform plan
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # keyhub_vaultrecord.record will be created
  + resource "keyhub_vaultrecord" "record" {
      + groupuuid = "4c25e8df-9bba-4e63-886c-b173d62fac55"
      + id        = (known after apply)
      + name      = "Terraform Test record"
      + password  = (sensitive value)
      + username  = "a_username_s"
      + uuid      = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

```
% terraform import keyhub_vaultrecord.record "7dd69a18-212e-4132-b487-1b21e37e7813"
keyhub_vaultrecord.record: Importing from ID "7dd69a18-212e-4132-b487-1b21e37e7813"...
keyhub_vaultrecord.record: Import prepared!
  Prepared keyhub_vaultrecord for import
keyhub_vaultrecord.record: Refreshing state... [id=7dd69a18-212e-4132-b487-1b21e37e7813]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.
```

```
% terraform plan
keyhub_vaultrecord.record: Refreshing state... [id=47639110]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # keyhub_vaultrecord.record will be updated in-place
  ~ resource "keyhub_vaultrecord" "record" {
      - filename  = "43137751.png" -> null
        id        = "47639110"
        name      = "Terraform Test record"
        # (4 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

Reply:
Je ziet wel dat tijdens de import als ID de meegegeven waarde getoond wordt
`keyhub_vaultrecord.record: Refreshing state... [id=7dd69a18-212e-4132-b487-1b21e37e7813]`
Maar dat in de state wel het DB id opgeslagen is (bij de 2e plan)
`keyhub_vaultrecord.record: Refreshing state... [id=47639110]`

Kan de documentatie zo ff niet terug vinden, maar de documentatie zegt dat de waarde bij een import niet perse het ID hoeft te zijn, als het maar een unieke waarde is waarmee je de resource kunt vinden in je achterliggende api. En gezien uuid zichtbaarder is voor keyhub gebruikers dan een DB id, lijkt me die veel beter om voor imports te gebruiken

Reply:
Ah ik wist niet dat dit zo was. Ik dacht dat je altijd het ID moest opgeven. Prima zo dan.